runtime/compiler/env/j9methodServer.cpp

```diff
@@ -633,62 +612,31 @@ TR_ResolvedJ9JITServerMethod::getResolvedStaticMethod(TR::Compilation * comp, I_
+   TR_ResolvedMethodKey key = compInfoPT->getResolvedMethodKey(TR_ResolvedMethodType::Static, (TR_OpaqueClassBlock *) _ramClass, cpIndex);
+   if (compInfoPT->getCachedResolvedMethod(key, this, &resolvedMethod, unresolvedInCP))
       {
       if ((resolvedMethod == NULL) && unresolvedInCP)
          handleUnresolvedStaticMethodInCP(cpIndex, unresolvedInCP);
       return resolvedMethod;
       }
 
    _stream->write(JITServer::MessageType::ResolvedMethod_getResolvedStaticMethodAndMirror, _remoteMirror, cpIndex);
-   auto recv = _stream->read<J9Method *, TR_ResolvedJ9JITServerMethodInfo, bool>();
-   J9Method * ramMethod = std::get<0>(recv);
+   auto recv = _stream->read<TR_OpaqueMethodBlock *, TR_ResolvedJ9JITServerMethodInfo, bool>();
+   TR_OpaqueMethodBlock *ramMethod = std::get<0>(recv);
+   auto methodInfo = std::get<1>(recv);
    bool isUnresolvedInCP = std::get<2>(recv);
    if (unresolvedInCP)
       *unresolvedInCP = isUnresolvedInCP;
 
-   if (comp->compileRelocatableCode() && comp->getOption(TR_UseSymbolValidationManager) && ramMethod)
+   if (ramMethod == NULL) 
       {
-      if (!comp->getSymbolValidationManager()->addStaticMethodFromCPRecord((TR_OpaqueMethodBlock *)ramMethod, cp(), cpIndex))
-         ramMethod = NULL;
-      }
-
-   bool skipForDebugging = doResolveAtRuntime(ramMethod, cpIndex, comp);
-   if (isArchetypeSpecimen())
-      {
-      // ILGen macros currently must be resolved for correctness, or else they
-      // are not recognized and expanded.  If we have unresolved calls, we can't
-      // tell whether they're ilgen macros because the recognized-method system
-      // only works on resovled methods.
-      //
-      if (ramMethod)
-         skipForDebugging = false;
-      else
+      if (unresolvedInCP) 
          {
-         comp->failCompilation<TR::ILGenFailure>("Can't compile an archetype specimen with unresolved calls");
-         }
-      }
-
-   auto &methodInfo = std::get<1>(recv);
-   if (ramMethod && !skipForDebugging)
-      {
-      TR_AOTInliningStats *aotStats = NULL;
-      if (comp->getOption(TR_EnableAOTStats))
-         aotStats = & (((TR_JitPrivateConfig *)_fe->_jitConfig->privateConfig)->aotStats->staticMethods);
-      resolvedMethod = createResolvedMethodFromJ9Method(comp, cpIndex, 0, ramMethod, unresolvedInCP, aotStats, methodInfo);
-      if (unresolvedInCP)
-         *unresolvedInCP = false;
-      }
-
-   if (resolvedMethod == NULL)
-      {
-      if (unresolvedInCP)
          handleUnresolvedStaticMethodInCP(cpIndex, unresolvedInCP);
+         }
+      return NULL;
       }
-   else
-      {
-      compInfoPT->cacheResolvedMethod(compInfoPT->getResolvedMethodKey(TR_ResolvedMethodType::Static, (TR_OpaqueClassBlock *) _ramClass, cpIndex), (TR_OpaqueMethodBlock *) ramMethod, 0, methodInfo);
-      }
+   resolvedMethod = compInfoPT->cacheAndGetResolvedMethod(key, ramMethod, 0, methodInfo, this, &resolvedMethod, unresolvedInCP);
```

Comment:
`cacheAndResolveMethod` calls `getCachedResolvedMethod` which does some SVM validation. If that fails we return `unresolvedInCP==true`, but that may not have happened in the non-JITServer code.

Reply:
Wouldn't non-JITServer code be in the wrong in that case? If we failed to add SVM record, then the resolved method will not be created, so `unresolvedInCP` must always be set to true, right?

Reply:
I am not entirely sure, but I thought that it is possible have `unresolvedInCP==false` and still fail to create the `resolvedMethod` because of SVM. For instance in `getResolvedPossiblyPrivateVirtualMethod` we check whether the method is resolved in 
`  J9Method * ramMethod = (J9Method *)getVirtualMethod(_fe, cp(), cpIndex, &vTableOffset, unresolvedInCP);`
which may set `unresolvedinCP` to `false` and yet, due to SVM failing we set `createResolvedMethod = false` which will make us return `false` in the end.